assertion failure in H5D__mpio_collective_filtered_vec_io

**Describe the bug**
Crash in t_pmulti_dset test:

Configuration:
  Layout:           Chunked
  Shape same:       No
  I/O type:         Single
  MPI I/O type:     Collective
  Low level MPI I/O:Collective
  Type conversion:  No
  Data filter:      Yes
Testing random I/O                                                    Assertion failed: (vec_idx < iovec_count), function H5D__mpio_collective_filtered_vec_io, file H5Dmpio.c, line 5498.

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 91513 RUNNING AT Nile.local
=   EXIT CODE: 6
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Terminated: 15 (signal 15)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
        7.23 real        33.77 user         8.05 sys
make[3]: *** [t_pmulti_dset.chkexe_] Error 1
make[2]: *** [build-check-p] Error 1
make[1]: *** [test] Error 2
make: *** [check-am] Error 2


**Expected behavior**
A clear and concise description of what you expected to happen.

**Platform (please complete the following information)**
Current version of the develop branch, on my Mac, with this configuration:

env CC=mpicc CXX=mpic++ FC=mpif90 ../hdf5/configure --enable-build-mode=debug --enable-maintainer-mode --enable-developer-warnings --enable-diags --disable-shared --enable-parallel --enable-cxx --enable-fortran --enable-map-api --enable-threadsafe --enable-build-all --enable-mirror-vfd --enable-subfiling-vfd --enable-unsupported

**Additional context**
Add any other context about the problem here.

--- a/src/H5Dmpio.c
+++ b/src/H5Dmpio.c
@@ -5489,6 +5489,9 @@ H5D__mpio_collective_filtered_vec_io(const H5D_filtered_collective_io_info_t *ch
         for (size_t i = 0, vec_idx = 0; i < chunk_list->num_chunk_infos; i++) {
             H5F_block_t *chunk_block;
 
+            if (op_type == H5D_IO_OP_READ && !chunk_list->chunk_infos[i].need_read)
+                continue;
+
             /*
              * Check that we aren't going to accidentally try to write past the
              * allocated memory for the I/O vector buffers in case bookkeeping
@@ -5496,9 +5499,6 @@ H5D__mpio_collective_filtered_vec_io(const H5D_filtered_collective_io_info_t *ch
              * field.
              */
             assert(vec_idx < iovec_count);
-
-            if (op_type == H5D_IO_OP_READ && !chunk_list->chunk_infos[i].need_read)
-                continue;
 
             /* Set convenience pointer for current chunk block */
             chunk_block = (op_type == H5D_IO_OP_READ) ? &chunk_list->chunk_infos[i].chunk_current

